packages/rum-core/src/domain/contexts/pageStateContexts.ts

```diff
@@ -0,0 +1,141 @@
+import type { Duration, RelativeTime } from '@datadog/browser-core'
+import {
+  isExperimentalFeatureEnabled,
+  noop,
+  elapsed,
+  addDuration,
+  addEventListeners,
+  relativeNow,
+  DOM_EVENT,
+} from '@datadog/browser-core'
+
+export const PAGE_STATE_TIMELINE_MAX_LENGTH = 500
+
+export const enum PageState {
+  ACTIVE = 'active',
+  PASSIVE = 'passive',
+  HIDDEN = 'hidden',
+  FROZEN = 'frozen',
+  TERMINATED = 'terminated',
+}
+
+type PageStateTimeline = Array<{ state: PageState; startTime: RelativeTime }>
+
+type Counter = { count: number; duration: Duration }
+export type PageStateContext = {
+  active: Counter
+  passive: Counter
+  hidden: Counter
+  frozen: Counter
+  terminated: Counter
+}
+
+export interface PageStateContexts {
+  getPageStates: (startTime: RelativeTime, duration: Duration) => PageStateContext | undefined
+  stop: () => void
+}
+
+let pageStateTimeline: PageStateTimeline = []
+let state: PageState | undefined
+
+export function startPageStateContexts(): PageStateContexts {
+  if (!isExperimentalFeatureEnabled('resource_page_states')) {
+    return {
+      getPageStates: () => undefined,
+      stop: noop,
+    }
+  }
+
+  addPageState(getState())
+
+  const { stop } = addEventListeners(
+    window,
+    [
+      DOM_EVENT.PAGE_SHOW,
+      DOM_EVENT.FOCUS,
+      DOM_EVENT.BLUR,
+      DOM_EVENT.VISIBILITY_CHANGE,
+      DOM_EVENT.RESUME,
+      DOM_EVENT.FREEZE,
+      DOM_EVENT.PAGE_HIDE,
+    ],
+    (event) => {
+      if (!event.isTrusted) return
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏what is the intent of this `isTrusted` check?

Reply:
While testing the page state context locally, I saw a weird behaviour related to the fact that our chrome extension[ trigger custom  'visibilitychange' event ](https://github.com/DataDog/browser-sdk/blob/b604215075db0a483f3b6674e31c69a3fa83194b/developer-extension/src/panel/flushEvents.ts#L11). I want to prevent that.

Reply:
Oh, I guess that many states were added due to the auto flush.

it could be nice to add a comment explaining the intent.